ex5/r_kobayashi/k_mean.py

```diff
@@ -0,0 +1,343 @@
+        save_fname = os.path.join(path, "result", save_fname)
+        ani.save(save_fname, writer="pillow")
+    plt.tight_layout()
+    plt.show()
+
+
+def minimax(data, K):
+    """
+    ミニマックス法
+
+    paramerters
+    -----------
+    data : numpy.ndarray
+        csv data
+    K : int
+        the number of cluster
+      
+    return
+    ------
+    cen : numpy.ndarray
+        center of cluster
+    """
+    num, dim = data.shape
+    cidx = random.randint(0, num-1)
+    dis = np.zeros((K, num))
+    cen = np.zeros((K, dim))
+    for k in range(K):
+        cen[k] = data[cidx]
+        # 距離計算
+        r = np.sum((data-data[cidx])**2, axis=1)
+        # 距離保存
+        dis[k] =  r
+        cidx = np.argmax(np.min(dis[:k+1], axis=0))
+
+    return cen
+
+
+def kplus(data, K):
+    """
+    kmeans++
+
+    paramerters
+    -----------
+    data : numpy.ndarray
+        csv data
+    K : int
+        the number of cluster
+      
+    return
+    ------
+    cen : numpy.ndarray
+        center of cluster
+    """
+    num, dim = data.shape
+    cidx = random.randint(0, num-1)
+    dis = np.zeros((K, num))
+    cen = np.zeros((K, dim))
+    pr = np.zeros(num)
+    for k in range(K):
+        cen[k] = data[cidx]
+        # 距離計算
+        r = np.sum((data-data[cidx])**2, axis=1)
+        # 距離保存
+        dis[k] =  r
+        
+        # 確率作成
+        pr = np.min(dis[:k+1], axis=0)
+        pr = pr / np.sum(pr)
+        # 次の中心
+        cidx = np.random.choice(np.arange(num), 1, p=pr)
+
+    return cen
+
+
+def LGB(data, K, cen):
+    """
+    LGB法
+
+    paramerters
+    -----------
+    data : numpy.ndarray
+        csv data
+    K : int
+        the number of cluster
+    cen : numpy.ndarray
+        center of cluster
+    
+    return
+    ------
+    newcen : numpy.ndarray
+        new center of cluster
+    """
+    # クラスタの中心をふたつに分ける
+    delta = 0.01
+    cen_b = cen-delta
+    cen_a = cen+delta
+    newcen = np.concatenate((cen_b, cen_a))
+    M = newcen.shape[0]
+    # kmeansアルゴリズム
+    newcen, clu = kmean(data, M, newcen)
+    if newcen.shape[0] >= K:
+        newcen = newcen[random.sample(range(len(newcen)), K)]
+        return newcen
+    else:
+        return LGB(data, K, newcen)
+        
+
+def method(data, K, mname):
+    """
+    初期値決定方法
+
+    paramerters
+    -----------
+    data : numpy.ndarray
+        csv data
+    K : int
+        the number of cluster
+    mname : str
+        method name
+      
+    return
+    ------
+    cen :   numpy.ndarray
+        center of cluster
+    """
+    if mname == "minimax":
+        cen = minimax(data, K)
+    elif mname == "kplus":
+        cen = kplus(data, K)
+    elif mname == "LGB":
+        if data.shape[1] == 2:
+            ce = np.zeros((1, 2))
+        elif data.shape[1] == 3:
+            ce = np.zeros((1, 3))
+        ce[0] = np.mean(data, axis=0)
+        cen = LGB(data, K, ce)
+    else:
+        print("error:method name")
+        sys.exit(1)

```

Comment:
例外処理いいね👍
assert使うといいよ
参考：https://qiita.com/nannoki/items/15004992b6bb5637a9cd
```suggestion
    # 先に例外処理
    assert mname in ["minimax", "kplus", "LGB"], "Comment here"
    if mname == "minimax":
        # 処理を書く
    elif mname == "kplus":
        # 処理を書く
    elif mname == "LGB":
        # 処理を書く
```